Enhance fakepg: add fsdp+tp tests

xmfan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107626
* #107625

from working on a starter task with @wanchaol (T161350434):
Add the previously unsupported fsdp+tp example in FakePG, which required scatter and broadcast, as a unit test: https://github.com/pytorch/examples/blob/main/distributed/tensor_parallelism/two_d_parallel_example.py

xmfan committed:
Enhance fakepg: add fsdp+tp tests

[ghstack-poisoned]

fduwjj commented on test/distributed/test_fake_pg.py:
```diff
@@ -168,6 +175,73 @@ def test_recv(self):
+            "cuda", torch.arange(0, world_size).view(-1, tp_size)
+        )
+
+        my_module = nn.Sequential(
+            nn.Linear(10, 32),
+            nn.ReLU(),
+            nn.Linear(32, 5),
+        ).cuda(dist.get_rank())
```
Can we reuse the module defined here: https://github.com/pytorch/pytorch/blob/9c56ca80f36f98cb01190eb63053b07c15e95038/torch/testing/_internal/distributed/_tensor/common_dtensor.py#L53. Or if you don't want to use that one, can we make this more reusable so that we avoid duplicate code? Thanks!

fduwjj commented on test/distributed/test_fake_pg.py:
```diff
@@ -168,6 +175,73 @@ def test_recv(self):
+            optim.step()
+
+    @unittest.skipIf(not HAS_CUDA or not enable_2d_with_fsdp(), "No CUDA or TP+FSDP")
+    def test_fsdp_tp_pp_fake_e2e(self):
```
ditto, this one is very similar to the SP one, can we consolidate them a little bit? Thanks!

wanchaol commented on test/distributed/test_fake_pg.py:
```diff
@@ -168,6 +175,73 @@ def test_recv(self):
+            optim.step()
+
+    @unittest.skipIf(not HAS_CUDA or not enable_2d_with_fsdp(), "No CUDA or TP+FSDP")
+    def test_fsdp_tp_pp_fake_e2e(self):
```
naming seems indicating we have "pp" but the test does not have pipeline parallel? Agree with @fduwjj that we should consolidate the SP/TP test case a bit, the only difference here looks to be the `parallel_style`, we can simply loop through parallel styles and verify each case works in a single test case

xmfan replied:
pp was intended to refer to using `PairwiseParallel` instead of `SequenceParallel`, I'll change it since it's confusing

The other minor difference with SP test case is how we retrieve the `dp_rank`, but it can be combined into a loop as well

xmfan committed:
Update on "Enhance fakepg: add fsdp+tp tests"


from working on a starter task with wanchaol (T161350434):
Add the previously unsupported fsdp+tp example in FakePG, which required scatter and broadcast, as a unit test: https://github.com/pytorch/examples/blob/main/distributed/tensor_parallelism/two_d_parallel_example.py

[ghstack-poisoned]

xmfan commented:
@xmfan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48606652).

wanchaol reviewed (approved):
lgtm, thanks for addressing comments!

xmfan committed:
Update on "Enhance fakepg: add fsdp+tp tests"


from working on a starter task with wanchaol (T161350434):
Add the previously unsupported fsdp+tp example in FakePG, which required scatter and broadcast, as a unit test: https://github.com/pytorch/examples/blob/main/distributed/tensor_parallelism/two_d_parallel_example.py

Differential Revision: [D48606652](https://our.internmc.facebook.com/intern/diff/D48606652)

[ghstack-poisoned]

xmfan commented:
@pytorchbot merge
